### PR TITLE
Article/monorepos

### PR DESCRIPTION
**Pull request by matt-tyler**
The article details some of the work taken to manage the ever expanding build times of the mono-repo in CNF.

There is still additional work to add some concrete technical examples, and add a conclusion, but the bulk of the story is there.

### CODE REVIEW

**Comment by matt-tyler** [comment]
Additionally, I might post this one to my personal blog, medium, and dev.to and set the rel=canonical links back to the mech rock version of the article. That might help to boost traffic towards the MR blog/site.

Do we have any analytics on the site currently so we can try and measure any effect?

File: _posts/2019-01-18-monorepos-aws-codebuild.markdown [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+The first step was to introduce something to help us manage the monorepo. I had previously used 'lerna' to manage another node-based monorepo with some success. This began a reasonably lengthy task to restructure what he already had. This meant moving from a heavily nested module structure, to a more flattened structure - which is technically what would have happened had each module been seperated into its own repository. With this in-place, it was now a little easier to manage the dependencies of each module and enable a clear seperation of concerns. We were also using a feature in lerna called 'hoisting' which deduplicates dependencies that many packages may rely on.
```
Comment (amitsha) [inline _posts/2019-01-18-monorepos-aws-codebuild.markdown]: `Thus` began ... what `we` already had.

File: _posts/2019-01-18-monorepos-aws-codebuild.markdown [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+Despite the movement to lerna to better manage the repository, this meant that we still had to reckon with fairly large build times which were agitated by the removal of hoisting. It was at this point I started inspecting our build environment and actions to determine possible improvements. This involved learning a lot more as to how npm packaging, caching and installation works, and from this I was able to come up with a few things to try out.
+
+The next improvement to be made was to use `npm ci`. This instructs npm to not resolve dependencies from directly from definitions in the package.json file, and instead use the package-lock.json file. Fundamentally, this instructs npm to install all packages called out in the lock file, rather than resolving everything from the top level dependencies. This provided a reasonably modest speed improvement (roughly ~20%), which is not bad for what was a minor one-line change.
```
Comment (amitsha) [inline _posts/2019-01-18-monorepos-aws-codebuild.markdown]: This instructs npm to not resolve dependencies from directly from definitions in the package.json file -> 
extra from

File: _posts/2019-01-18-monorepos-aws-codebuild.markdown [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+Some additional changes were needed to make Jest perform slightly better in this arrangement. Previously, we called jest seperately on each project, with each project having its own seperate jest configuration. We instead provided a global jest configuration at the base of the monorepo that was capable of locating and executing all tests across the entire repository. This does require that you name and locate tests based upon a convention, which fortunately we were doing.
+
+At this point, we had solved the worst of our build time issues, resulting in a reduction from approximately 15 minutes down to 3 minutes for a complete build of the repository. 
+
+There are still additional optimizations that can be made. We currently have a configuration to use Jest in multi-project mode, which when combined with lerna's 'changed' command, can be used to ensure we that only build and test packages in the repository that have changed. This would make our CI check run much faster for changes that only touch a few packages (which has the added effect of encouraging our developers to make many smaller changes as opposed to fewer larger ones).
```
Comment (amitsha) [inline _posts/2019-01-18-monorepos-aws-codebuild.markdown]: 🥇

**Comment by matt-tyler** [comment]
This is pretty much read to go (I think). Just needs to have additional review and dates changed.

File: _posts/2019-01-18-monorepos-aws-codebuild.markdown [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+There are two ways to improve this - better hardware, and a reduction in the number of times these dependencies would installed. The former was achieved by bumping the size of the build environment. The latter was slightly more complex. We emulated the hoisting feature by moving development dependencies to top level package.json, and called out these dependencies as peer dependencies to serve as a reminder that they were required in the child packages. 
+
+Some additional changes were needed to make Jest perform slightly better in this arrangement. Previously, we called jest seperately on each project, with each project having its own seperate jest configuration. We instead provided a global jest configuration at the base of the monorepo that was capable of locating and executing all tests across the entire repository. This does require that you name and locate tests based upon a convention, which fortunately we were doing.
+
+At this point, we had solved the worst of our build time issues, resulting in a reduction from approximately 15 minutes down to 3 minutes for a complete build of the repository.
```
Comment (Hamo69) [inline _posts/2019-01-18-monorepos-aws-codebuild.markdown]: call this out / highlight it

File: _posts/2019-01-18-monorepos-aws-codebuild.markdown [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+This allows us to target execute tests against only packages which have changed. This does require you to perform a full checkout of the repository in AWS Codebuild, as opposed to the default behaviour which is to perform a shallow clone. We don't currently have this activated as we had been waiting for Jest 24 to release in order to resolve some other unrelated issues. With Jest 24 having been recently launch, I look forward to turning this on!
+
+I hope this gives everyone a good look at the steps that need to be taken to keep the build process of a monorepo running reasonably efficiently. In this case, it has included;
+
+- Installing specialised tooling to manage a monorepo, in our case it was lerna.
+- Taking advantage of dependency caching in AWS CodeBuild.
+- Utilising npm ci to resolve and install dependencies.
+- Running jest in multi-project mode together with lerna changed.
+
+Hopefully this has been helpful at anyone who is looking at taking the Monorepo approach.
```
Comment (Hamo69) [inline _posts/2019-01-18-monorepos-aws-codebuild.markdown]: maybe add your contact details or mech rocks, if there are anyone experiencing similar build issues and needs help ;-)

**Comment by matt-tyler** [comment]
@Hamo69 @amitsha added your changes - re-read and see if you're happy

**Pull request merged**